File: rp4_gpio/HX711.py
import threading

from RPduino import *

mux = threading.Lock()

class HX711():
    def __init__(self, DOUT: int, SCK: int):
        self.DOUT = DOUT
        self.SCK = SCK
        self.GAIN = 1
        self.OFFSET = 0
        self.RESULT = 0
        self.DESCRIPTION = ''

        pinMode(self.DOUT, INPUT)
        pinMode(self.SCK, OUTPUT)

    def get_result(self) -> int:
        return self.RESULT

    def get_offset(self) -> int:
        return self.OFFSET

    def set_result(self, new_RESULT: int):
        self.RESULT = new_RESULT - self.OFFSET

    def set_offset(self, new_OFFSET: int):
        self.OFFSET = new_OFFSET

    def set_gain(self, new_GAIN: int):
        '''
        :param new_GAIN: new GAIN 128 / 64 / 32
        '''
        match new_GAIN:
            case 128:
                self.GAIN = 1
            case 64:
                self.GAIN = 2
            case 32:
                self.GAIN = 3
            case _:
                raise ValueError("new_GAIN must be 128 / 64 / 32")            

def set_offset(hx711: HX711):
    hx711.set_offset(0)
    hx711.set_offset(read(hx711))

def is_ready(hx711: HX711) -> bool:
    return digitalRead(hx711.DOUT) == 0

def wait_ready(hx711: HX711, time: int):
    '''
    :param time: time in ms
    :type time: int
    '''
    while(not is_ready(hx711)): delay(time)

def shiftIn(dataPin: int, clockPin: int) -> int:

    value = 0

    for i in range(8):
        digitalWrite(clockPin, HIGH)
        value |= digitalRead(dataPin) << (7 - i)
        digitalWrite(clockPin, LOW)
        # Delay pre stabilizaciu
        # Bez toho moze citat falosne udaje
        delayMicros(1)

    return value

def read(hx711: HX711) -> int:
    wait_ready(hx711, 1)

    value = 0
    data = [0, 0, 0]

    with mux:
        for i in range(3):
            data[2 - i] = shiftIn(hx711.DOUT, hx711.SCK)

        for i in range(hx711.GAIN):
            digitalWrite(hx711.SCK, HIGH)
            delayMicros(1)
            digitalWrite(hx711.SCK, LOW)
            delayMicros(1)
    
    value = data[2] << 8*2 \
          | data[1] << 8   \
          | data[0]

    if value & 0x800000:
        value -= 0x1000000

    hx711.set_result(value)

    return value

File: ESP32/HX711.cpp
#include "HX711.h"

bool is_ready(struct HX711 *hx711){
    return digitalRead(hx711->DOUT) == 0;
}

void set_gain(struct HX711 *hx711, uint8_t new_gain){
    uint8_t gain = 1;
    switch (new_gain) {
        case 128:
            gain = 1;
            break;
        case 64:
            gain = 3;
            break;
        case 32:
            gain = 2;
            break;
    }

    hx711->GAIN = gain;
}

void wait_ready(struct HX711 *hx711, uint32_t time){
    while (!is_ready(hx711)) delay(time);
}

uint8_t shiftInSlow(uint8_t dataPin, uint8_t clockPin, uint8_t bitOrder) {
    uint8_t value = 0;
    uint8_t i;

    for (i = 0; i < 8; ++i) {
        digitalWrite(clockPin, HIGH);
        delayMicroseconds(1);
        if (bitOrder == LSBFIRST)   
            value |= digitalRead(dataPin) << i;
        else
            value |= digitalRead(dataPin) << (7 - i);
        digitalWrite(clockPin, LOW);
        delayMicroseconds(1);
    }
    return value;
}


int32_t read(struct HX711 *hx711){
    wait_ready(hx711, 1);

    uint32_t value = 0;
    uint8_t data[3] = { 0 };
    uint8_t filler = 0x00;

	portMUX_TYPE mux = portMUX_INITIALIZER_UNLOCKED;
	portENTER_CRITICAL(&mux);

    for (uint8_t i = 0; i < 3; i++)
        data[2 - i] = shiftInSlow(hx711->DOUT, hx711->SCK, MSBFIRST);

    for (uint8_t i = 0; i < hx711->GAIN; i++){
        digitalWrite(hx711->SCK, HIGH);
        delayMicroseconds(1);
        digitalWrite(hx711->SCK, LOW);
        delayMicroseconds(1);
    }

	portEXIT_CRITICAL(&mux);

    filler = data[2] & 0x80 ? 0xFF : 0x00;

    value = ((uint32_t)filler << 8*3)
          | ((uint32_t)data[2] << 8*2) 
          | ((uint32_t)data[1] << 8)
          | ((uint32_t)data[0]);

    return (int32_t)value;
}

int32_t read_avg(struct HX711 *hx711, uint8_t times){
    int32_t sum = 0;
    for (uint8_t i = 0; i < times; i++)
        sum += read(hx711);

    return sum;
}

void tare(struct HX711 *hx711, uint8_t times){
    int32_t sum = read_avg(hx711, times);
    // set_offset(sum);
}

// void set_offset(int32_t offset){
//     return;
// }